### PR TITLE
feat(arp): Implement job to expire stale POA requests (#106319)

### PR DESCRIPTION
**Pull request by ojbucao**
### Description

This PR introduces a background job to automatically expire Power of Attorney (POA) requests within the Accredited Representative Portal module. The job targets requests that remain unresolved after 60 days from their creation, helping manage the request lifecycle and ensuring stale data is handled appropriately.

---

### Changes

* **Added `AccreditedRepresentativePortal::ExpirePowerOfAttorneyRequestsJob`:**
    * A Sidekiq job designed to run periodically (e.g., nightly).
    * Queries for `PowerOfAttorneyRequest` records that are `.unresolved` and older than `PowerOfAttorneyRequest::EXPIRY_DURATION` (60 days), correctly handling potential column ambiguity in the SQL query.
    * Resolves eligible requests by creating an associated `PowerOfAttorneyRequestExpiration` record via `PowerOfAttorneyRequestResolution` polymorphic association using the custom `create_with_resolving!` method.
    * Includes error handling to log failures for individual requests while allowing the job to continue processing others.
    * Refactored into smaller private methods for better readability and RuboCop compliance (`Metrics/MethodLength`).
* **Added RSpec Tests (`expire_power_of_attorney_requests_job_spec.rb`):**
    * Provides comprehensive test coverage for the new job.
    * Verifies correct identification of eligible/ineligible requests based on age and resolution status.
    * Confirms the creation of the correct expiration resolution type.
    * Tests the error handling mechanism.
* **Updated `PowerOfAttorneyRequest` Model:**
    * Added `dependent: :destroy` to the `has_one :power_of_attorney_form` association. This is crucial to prevent `ActiveRecord::InvalidForeignKey` (specifically `PG::ForeignKeyViolation`) errors during test database cleanup (and potentially other scenarios where requests are deleted) by ensuring the associated form record is deleted first.
* **Updated `PowerOfAttorneyRequest` Factory:**
    * Added an explicit `:unresolved` trait for improved clarity in test setups, even though it represents the default state.

---

### Testing Done

* Added extensive unit/integration tests for the `ExpirePowerOfAttorneyRequestsJob` using RSpec and `Sidekiq::Testing`.
* All tests pass locally, covering the core logic, edge cases (boundary time), interaction with already-resolved requests, and error handling.
* Verified the `dependent: :destroy` fix resolves the previously encountered foreign key violations during test runs.

## Related issue(s)
https://github.com/department-of-veterans-affairs/va.gov-team/issues/105142
https://github.com/department-of-veterans-affairs/va.gov-team/issues/106319

## Testing done

- [x] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      log_expiration_error(request, e)
+      false # Indicate failure
+    end
+
+    # Performs the actual expiration logic for a single request
+    def expire_request(request)
+      log_request_expiration(request)
+
+      expiration = PowerOfAttorneyRequestExpiration.new
+      PowerOfAttorneyRequestResolution.create_with_resolving!(
```
Comment (nihil2501) [inline modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:65]: `PowerOfAttorneyRequestExpiration.create_with_resolution!` would be a convenient helper mirroring the others.

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/power_of_attorney_request.rb [hunk 2]
```diff
@@ -8,7 +8,8 @@ class PowerOfAttorneyRequest < ApplicationRecord
 
     has_one :power_of_attorney_form,
             inverse_of: :power_of_attorney_request,
-            required: true # for now
+            required: true, # for now
+            dependent: :destroy
```
Comment (nihil2501) [inline modules/accredited_representative_portal/app/models/accredited_representative_portal/power_of_attorney_request.rb:12]: Does this cause ActiveRecord to destroy POA forms when POA requests are destroyed? If so,
- We won't be deleting POA requests
- Relying on models' `destroy` which runs all of AR's lifecycle callback code and such is a pretty heavy approach

File: modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      PowerOfAttorneyRequest
+        .unresolved
+        .where("#{PowerOfAttorneyRequest.table_name}.created_at < ?", threshold)
+    end
+
+    def process_requests(requests_to_process)
+      results = { expired: 0, errors: 0 }
+
+      # Use find_each to process in batches, avoiding high memory usage.
+      requests_to_process.find_each do |request|
```
Comment (nihil2501) [inline modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:39]: Nice. Glad they have this convenience.

**Review by nihil2501 — DISMISSED** [review]
Looks good. A couple small suggestions.

We should consider possibly doing the redaction in-line too. I think ARM is working on adding a claimant expiration notification in-line here. And technically there'd be a very narrow race condition otherwise.

File: modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    private
+
+    def calculate_expiration_threshold
+      Time.current - PowerOfAttorneyRequest::EXPIRY_DURATION
+    end
+
+    def find_eligible_requests(threshold)
+      PowerOfAttorneyRequest
+        .unresolved
+        .where("#{PowerOfAttorneyRequest.table_name}.created_at < ?", threshold)
```
[the remark this answers is not in the document]
Comment (ojbucao) [reply modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:32]: Not specifying results in the PG::AmbiguousColumn error.

[the remark this answers is not in the document]
Comment (nihil2501) [reply modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:32]: But the snipped and output I pasted shows that the framework specifies the table automatically, no?

**Review by nihil2501 — APPROVED** [review]
LGTM

Nit: `create_expiration!` is redundant compared to `create_with_resolution!`—maybe sneak that tweak into one of the open PRs that is still waiting on review from the team.

File: modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb
Comment (rmtolmach) [inline modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:1]: You probably already know about it, but I wanted to pass on this [Sidekiq Job Standards](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/teams/benefits/architectural-decision-records/sidekiq-async-guidance.md) doc. I just want to make sure people writing sidekiq jobs are aware of it! 🙂

File: modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+require 'sidekiq'
+
+module AccreditedRepresentativePortal
+  # Job to find and expire PowerOfAttorneyRequests that are older than
+  # EXPIRY_DURATION (60 days) and remain unresolved.
+  class ExpirePowerOfAttorneyRequestsJob
```
[the remark this answers is not in the document]
Comment (ojbucao) [reply modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:8]: Hi @rmtolmach. We'll add it to periodic_jobs in a separate PR.

[the remark this answers is not in the document]
Comment (rmtolmach) [reply modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/expire_power_of_attorney_requests_job.rb:8]: ok 👍

**Review by rmtolmach — APPROVED** [review]
See my comment about making sure to squash and merge ☝️

**Comment by ojbucao** [comment]
Oof you're absolutely right to call that out! 

Thanks for the reminder!

**Review by rmtolmach — APPROVED** [review]
re-approving

**Comment by rmtolmach** [comment]
🎉 Squashing and merging made this show up in one commit. thanks!
image
